File: src/main/java/dev/kensa/util/Strings.java
package dev.kensa.util;

import java.util.regex.Pattern;

public final class Strings {

    private static Pattern CAMEL_SPLIT_PATTERN = Pattern.compile(
            String.format(
                    "%s|%s|%s",
                    "(?<=[A-Z])(?=[A-Z][a-z])",     // Uppercase behind, Uppercase then Lowercase ahead
                    "(?<=[^A-Z])(?=[A-Z])",         // Non-Uppercase behind, Uppercase ahead
                    "(?<=[A-Za-z])(?=[0-9])"        // Letter behind, number ahead
            )
    );

    public static String unCamel(String phrase) {
        if (isBlank(phrase)) {
            return phrase;
        }

        phrase = phrase.replaceAll("[^a-zA-Z0-9]", "");

        String[] elements = CAMEL_SPLIT_PATTERN.split(phrase);
        elements[0] = Character.toUpperCase(elements[0].charAt(0)) + elements[0].substring(1);

        return String.join(" ", elements);
    }

    public static boolean isNotBlank(String value) {
        return !isBlank(value);
    }

    public static boolean isBlank(String value) {
        if (value == null || value.length() == 0) {
            return true;
        }

        return !value.chars()
                     .filter(c -> !Character.isWhitespace(c))
                     .findFirst()
                     .isPresent();
    }

    private Strings() {
    }
}

File: src/main/java/dev/kensa/util/Strings.kt
package dev.kensa.util

object Strings {
    private val CAMEL_SPLIT_REGEX = Regex("(?<=[A-Z])(?=[A-Z][a-z])|(?<=[^A-Z])(?=[A-Z])|(?<=[A-Za-z])(?=[0-9])")
    private val UPPER_CASE_LETTER_REGEX = Regex("(?=\\p{Lu})")

    fun unCamel(phrase: String): String {
        if (phrase.isBlank()) {
            return phrase
        }
        val replaced = phrase.replace("[^a-zA-Z0-9]".toRegex(), "")
        return CAMEL_SPLIT_REGEX.split(replaced).toTypedArray().let { elements ->
            elements[0] = Character.toUpperCase(elements[0][0]).toString() + elements[0].substring(1)
            elements.joinToString(" ")
        }
    }

    fun unCamelToSeparated(phrase: String, separator: String = "-"): String? {
        return if (phrase.isBlank()) {
            phrase
        } else {
            UPPER_CASE_LETTER_REGEX.split(phrase.trim()).filterNot { it.isBlank() }.joinToString(separator).toLowerCase()
        }
    }
}